pandas/core/indexes/multi.py

```diff
@@ -50,6 +50,101 @@
          target_klass='MultiIndex or list of tuples'))
 
 
+class BaseMultiIndexCodesEngine(object):
```

Comment:
If cython multiple inheritance is a problem, this can still go in cython and be a `cdef class` and the subclasses below can still inherit from it.

Reply:
I'm afraid I'm not following you: I have no problem in having the subclasses below inherit from this, but rather of _also_ inheriting from their base classes.

The only way out I see is a template analogous to ``index_class_helper.pxi.in``... but I really don't think it is worth the added complexity.

(My preference would be to not cythonize _anyway_, because I doubt any measurable gain can be attained, given that the cost overwhelmingly comes from numpy and ``Index.get_loc()`` calls - but I'm ready to be convinced I'm wrong)

Reply:
OK, now I understand, you're suggesting to keep the two subclasses in Python. Makes sense.

Reply:
... well, except that
- apart from ``cdef class``, the only internal method which could in principle be a ``cdef`` (``_codes_to_ints``) is precisely the one which would remain in ``multi.py`` (because it's type-specific)
- I would have to import ``pandas.core.dtypes.missing.isna`` inside ``index.pyx`` (maybe harmless, but I see all other current imports are lower level)

... but if you think this second point is not a problem, I will give it a try.